File: Source/Examples/ExampleLibraryCSharp/Sorting/SorterExamples.cs
/*  
  Copyright 2007-2013 The NGenerics Team
 (https://github.com/ngenerics/ngenerics/wiki/Team)

 This program is licensed under the GNU Lesser General Public License (LGPL).  You should 
 have received a copy of the license along with the source code.  If not, an online copy
 of the license can be found at http://www.gnu.org/copyleft/lesser.html.
*/

using System.Collections.Generic;
using NGenerics.Sorting;
using NUnit.Framework;

namespace ExampleLibraryCSharp.Sorting
{
    [TestFixture]
    public class SorterExamples
    {
        #region SortListComparer
        [Test]
        public void SortListComparerExample()
        {
            ComparisonSorter<int> sorter = new QuickSorter<int>();

            var list = new List<int> {13, 5, 77, 9, 12};

            sorter.Sort(list, Comparer<int>.Default);

            Assert.AreEqual(5, list[0]);
            Assert.AreEqual(9, list[1]);
            Assert.AreEqual(12, list[2]);
            Assert.AreEqual(13, list[3]);
            Assert.AreEqual(77, list[4]);
        }
        #endregion

        #region SortListDelegate
        [Test]
        public void SortListDelegateExample()
        {
            ComparisonSorter<int> sorter = new QuickSorter<int>();

            var list = new List<int> {13, 5, 77, 9, 12};

            sorter.Sort(list, IntComparison);

            Assert.AreEqual(5, list[0]);
            Assert.AreEqual(9, list[1]);
            Assert.AreEqual(12, list[2]);
            Assert.AreEqual(13, list[3]);
            Assert.AreEqual(77, list[4]);
        }


        private static int IntComparison(int i, int j)
        {
            if (i < j)
            {
                return -1;
            }
            if (i > j)
            {
                return 1;
            }
            return 0;
        }
        #endregion
        
        #region SortList
        [Test]
        public void SortListExample()
        {
            ComparisonSorter<int> sorter = new QuickSorter<int>();

            var list = new List<int> {13, 5, 77, 9, 12};

            sorter.Sort(list);

            Assert.AreEqual(5, list[0]);
            Assert.AreEqual(9, list[1]);
            Assert.AreEqual(12, list[2]);
            Assert.AreEqual(13, list[3]);
            Assert.AreEqual(77, list[4]);
        }
        #endregion

        #region SortListOrder
        [Test]
        public void SortListOrderExample()
        {
            ComparisonSorter<int> sorter = new QuickSorter<int>();

            var list = new List<int> {13, 5, 77, 9, 12};

            sorter.Sort(list, SortOrder.Ascending);

            Assert.AreEqual(77, list[4]);
            Assert.AreEqual(13, list[3]);
            Assert.AreEqual(12, list[2]);
            Assert.AreEqual(9, list[1]);
            Assert.AreEqual(5, list[0]);
        }
        #endregion
    }
}

File: Source/Examples/ExampleLibraryVB/Sorting/SorterExamples.vb
'  
' Copyright 2007-2013 The NGenerics Team
' (https://github.com/ngenerics/ngenerics/wiki/Team)
'
' This program is licensed under the GNU Lesser General Public License (LGPL).  You should 
' have received a copy of the license along with the source code.  If not, an online copy
' of the license can be found at http://www.gnu.org/copyleft/lesser.html.
'

Imports System
Imports System.Collections.Generic
Imports NGenerics.Sorting
Imports NUnit.Framework

<TestFixture()> _
Public Class SorterExamples
#Region "SortListComparer"
  <Test()> _
  Public Sub SortListComparerExample()
    Dim sorter As IComparisonSorter(Of Integer) = New QuickSorter(Of Integer)

    Dim list As New List(Of Integer)
    list.Add(13)
    list.Add(5)
    list.Add(77)
    list.Add(9)
    list.Add(12)

    sorter.Sort(list, Comparer(Of Integer).Default)

    Assert.AreEqual(5, list.Item(0))
    Assert.AreEqual(9, list.Item(1))
    Assert.AreEqual(12, list.Item(2))
    Assert.AreEqual(13, list.Item(3))
    Assert.AreEqual(77, list.Item(4))
  End Sub
#End Region

#Region "SortListDelegate"
  <Test()> _
  Public Sub SortListDelegateExample()
    Dim sorter As IComparisonSorter(Of Integer) = New QuickSorter(Of Integer)

    Dim list As New List(Of Integer)
    list.Add(13)
    list.Add(5)
    list.Add(77)
    list.Add(9)
    list.Add(12)

    sorter.Sort(list, New Comparison(Of Integer)(AddressOf SorterExamples.IntComparison))

    Assert.AreEqual(5, list.Item(0))
    Assert.AreEqual(9, list.Item(1))
    Assert.AreEqual(12, list.Item(2))
    Assert.AreEqual(13, list.Item(3))
    Assert.AreEqual(77, list.Item(4))
  End Sub

  Private Shared Function IntComparison(ByVal i As Integer, ByVal j As Integer) As Integer
    If (i < j) Then
      Return -1
    End If
    If (i > j) Then
      Return 1
    End If
    Return 0
  End Function
#End Region

#Region "SortList"
  <Test()> _
  Public Sub SortListExample()
    Dim sorter As Sorter(Of Integer) = New QuickSorter(Of Integer)

    Dim list As New List(Of Integer)
    list.Add(13)
    list.Add(5)
    list.Add(77)
    list.Add(9)
    list.Add(12)

    sorter.Sort(list)

    Assert.AreEqual(5, list.Item(0))
    Assert.AreEqual(9, list.Item(1))
    Assert.AreEqual(12, list.Item(2))
    Assert.AreEqual(13, list.Item(3))
    Assert.AreEqual(77, list.Item(4))
  End Sub
#End Region

#Region "SortListOrder"
  <Test()> _
  Public Sub SortListOrderExample()
    Dim sorter As Sorter(Of Integer) = New QuickSorter(Of Integer)

    Dim list As New List(Of Integer)
    list.Add(13)
    list.Add(5)
    list.Add(77)
    list.Add(9)
    list.Add(12)

    sorter.Sort(list, SortOrder.Ascending)

    Assert.AreEqual(77, list.Item(4))
    Assert.AreEqual(13, list.Item(3))
    Assert.AreEqual(12, list.Item(2))
    Assert.AreEqual(9, list.Item(1))
    Assert.AreEqual(5, list.Item(0))
  End Sub
#End Region
End Class
